### PR TITLE
plugin/etcdv3: Add etcd v3 plugin

### LINKED ISSUE #341: middleware/etcd: etcd3 support
See https://github.com/skynetservices/skydns/pull/299

### PR DESCRIPTION
**Pull request by nitisht**
### 1. What does this pull request do?
Add etcd v3 plugin

### 2. Which issues (if any) are related?
Fixes #341

### CODE REVIEW

File: plugin/etcdv3/README.md
Comment (harshavardhana) [inline plugin/etcdv3/README.md]: Remove this message

File: plugin.cfg
Comment (harshavardhana) [inline plugin.cfg]: better to call this etcd3 instead of etcdv3 while the latter is technically correct

File: plugin/etcdv3/etcdv3.go
Comment (harshavardhana) [inline plugin/etcdv3/etcdv3.go]: constants should be at the beginning of the file..

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP plugin/et..." ]
>+const (
>
>constants should be at the beginning of the file..

There is no such rule AFAICT. At the bottom is fine (doing this myself as well - 
quite like it)

**Comment by miekg** [comment]
you may need to tmp. add `go get github.com/coreos/etcd/clientv3`  to the make file so that we can build it in travis and friends (the godep update to fix the vendor dir can (and should) come later)

**Comment by nitisht** [comment]
> you may need to tmp. add go get github.com/coreos/etcd/clientv3 to the make file so that we can build it in travis and friends (the godep update to fix the vendor dir can (and should) come later)

Thanks, will do that @miekg

**Comment by miekg** [comment]
testing that makefile fix right now, CI barfs with
~~~
o test -v ./test/kubernetes/...
panic: http: multiple registrations for /debug/requests

goroutine 1 [running]:
net/http.(*ServeMux).Handle(0x28f91a0, 0x1bea346, 0xf, 0x285c720, 0x1c7c640)
	/usr/local/go/src/net/http/server.go:2270 +0x627
net/http.(*ServeMux).HandleFunc(0x28f91a0, 0x1bea346, 0xf, 0x1c7c640)
	/usr/local/go/src/net/http/server.go:2302 +0x55
net/http.HandleFunc(0x1bea346, 0xf, 0x1c7c640)
	/usr/local/go/src/net/http/server.go:2314 +0x4b
github.com/coreos/etcd/vendor/golang.org/x/net/trace.init.0()
	/tmp/tmp.AQ3DeOEBYO/src/github.com/coreos/etcd/vendor/golang.org/x/net/trace/trace.go:115 +0x42
github.com/coreos/etcd/vendor/golang.org/x/net/trace.init()
	:1 +0x1cd
github.com/coreos/etcd/vendor/google.golang.org/grpc.init()
	:1 +0x96
github.com/coreos/etcd/etcdserver/etcdserverpb.init()
	:1 +0x87
github.com/coreos/etcd/clientv3.init()
	:1 +0x5f
github.com/coredns/coredns/plugin/etcd3.init()
	:1 +0x7d
github.com/coredns/coredns/core/plugin.init()
	:1 +0x7f
github.com/coredns/ci/test/kubernetes.init()
	:1 +0x91
main.init()
	:1 +0x53
exit status 2
FAIL	github.com/coredns/ci/test/kubernetes	0.009s
Makefile:74: recipe for target 'test-k8s' failed
make: *** [test-k8s] Error 1
~~~

**Comment by miekg** [comment]
oh shit, looks like you can't have etcd2 and etcd3 at the same time in the same binary - at least not how we do it

**Comment by miekg** [comment]
yes, removing etcd makes it compile

**Comment by miekg** [comment]
Interestingly we now (probably) need to modify etcd too to be able to merge this - not sure how that looks - we can also *delete* etcd and make this the default - which is fine, but means we need to support the old way of configuring etcd - which in turn means we may just as well overwrite etcd entirely and make this the new default.

The only downside to this is when something is using etcd2 and can't possibly upgrade to etcd3 - which I can't really imagine.

**Comment by harshavardhana** [comment]
> There is no such rule AFAICT. At the bottom is fine (doing this myself as well -
> quite like it)

constants are like comments in the code sort of verbatim, it is better to define them at start of a file you can look at Go stdlib codebase for this approach. Not that its a rule but a good practice. I have never seen them added at the end of files which is quite confusing when navigating in editors.

**Comment by nitisht** [comment]
> which in turn means we may just as well overwrite etcd entirely and make this the new default.

Shall I go ahead with this, then @miekg ?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP plugin/et..." ]
>> which in turn means we may just as well overwrite etcd entirely and make this the new default.
>
>Shall I go ahead with this, then @miekg ?

Yes, I see etcd3 is the better plugin and CoreOS also says (for years) that 
everyone should move... @m4r10k : you're a heavy user of the etcd plugin, what 
do you think? Specically: we keep the Corefile syntax exactly the same, but 
use the etcd3 protocol, instead of the v2 one?

/Miek

--
Miek Gieben

**Comment by m4r10k** [comment]
> you're a heavy user of the etcd plugin, what
do you think? Specically: we keep the Corefile syntax exactly the same, but
use the etcd3 protocol, instead of the v2 one?

I've read through this PR and if the Corefile syntax stays the same, I see no problem in the change to the etcd v3 protocol the hard way, beside the migration of the entries from etcd v2 -> to etcd v3 for the inexperienced users.

Some background: We have etcd v3 clusters for kubernetes running (all on-prem) and the etcd v3 clusters can also be used with the etcd v2 protocol. So yes, we are storing the key/vales for coredns in the etcd v2 tree of the etcd v3 version currently. Using the etcd v3 tree won't be a great problem. Kubernetes support of etcd v2 as backend is [deprecated with kubernetes v 1.13 too](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.10.md) - in roughly 18 month I think.

Another benefit for us is, that the backup functionality of etcd v3 is much more usable than the one of etcd v2 for sure. So :+1: from my point of view.

**Comment by miekg** [comment]
What do you mean with "migrating", I was under the impression you could
just flip the transport and the etcd as-is?

**Comment by m4r10k** [comment]
@miekg I am not sure but if I use etcdctl to get key/values written with the v2 than they are not visible if they are asked by a v3 protocol query:

```
/ # etcdctl --cert-file /certs/coredns.pem --key-file /certs/coredns-key.pem --ca-file /certs/ca.pem --endpoint https://x.x.x.x:2379 get /dns-internal/com/aaa/bbb/ccc
{"host":"10.x.x.x","ttl":60}
/ # ETCDCTL_API=3 etcdctl --cert /certs/coredns.pem --key /certs/coredns-key.pem --cacert /certs/ca.pem --endpoints  https://x.x.x.x:2379 ge
t /dns-internal/com/aaa/bbb/ccc
```

The same is true vice versa:
```
/ # ETCDCTL_API=3 etcdctl --cert /certs/coredns.pem --key /certs/coredns-key.pem --cacert /certs/ca.pem --endpoints h https://x.x.x.x:2379 put /foo bar
OK
/ # ETCDCTL_API=3 etcdctl --cert /certs/coredns.pem --key /certs/coredns-key.pem --cacert /certs/ca.pem --endpoints h https://x.x.x.x:2379 get /foo
/foo
bar
/ # etcdctl --cert-file /certs/coredns.pem --key-file /certs/coredns-key.pem --ca-file /certs/ca.pem --endpoint ht https://x.x.x.x:2379 get /foo
Error:  100: Key not found (/foo) [132
```
I have not tested it with the etcd golang clientv3 library yet, to produce evidence if it is the case then too. If it is true, than someone have to migrate the entries from v2 to v3. This might be a special case for the etcdctl binary but not for the library - I do not know. Maybe I can test this later...

**Comment by miekg** [comment]
Damn, then doing a switcheroo is not backwards compatible.

@nitisht how hard would the following be to implement:
make etcd3 the new etcd and add a directive `version VERSION` that *defaults* to version 2, but can take a `3` as well; for 2 we use etcdv2 for 3 we use etcdv3 - this is backwards compatible and gives people to upgrade to 3 on their leisure.

**Comment by m4r10k** [comment]
> oh shit, looks like you can't have etcd2 and etcd3 at the same time in the same binary - at least not how we do it

I have not looked in this and I didn't tried it until yet, but I looked into https://github.com/coreos/etcd/tree/master/etcdctl briefly. I am not that coding guy, but I guess they made modules to separate the v2 code from the v3 code (ctlv2/ctlv3) and use accordingly [here](https://github.com/coreos/etcd/blob/b933603a4331054d0215819666b350ac3bba99f4/etcdctl/main.go#L30). 

So an option might be to use `etcd` as v2 plugin and `etcdv3` as v3 plugin and start from scratch there? Then it would be not a Version switch (additional option to the existing plugin) but a plugin switch which would be more "clear"?

**Comment by nitisht** [comment]
> how hard would the following be to implement:

Should not be too difficult @miekg . I will take a dig at it during the coming weekend.

**Comment by nchrisdk** [comment]
@nitisht any news on the progress of this PR?

**Comment by nitisht** [comment]
been busy with some work stuff, but will update it in a couple of days @nchrisdk

**Comment by miekg** [comment]
while talking with @m4r10k today at KubeCon I got the idea to go full on with etcd and allow all RR types in etcd instead of the subset we have now.

It does intersect with the sql work as well. I might abstract away the backend in the *file* plugin and see how hard it actually is (encoding RR's is probably the hard part)

**Comment by johnbelamaric** [comment]
Also consider that we may want both a *read* backend interface and a *write* backend interface, ala #522. Plugins could implement one or both. Also then we can build APIs on top of the write interface - DDNS, gRPC, REST.

**Comment by nchrisdk** [comment]
@nitisht I can understand that. Let me know if I can help.

**Comment by nitisht** [comment]
> make etcd3 the new etcd and add a directive version VERSION that defaults to version 2, but can take a 3 as well; for 2 we use etcdv2 for 3 we use etcdv3 - this is backwards compatible and gives people to upgrade to 3 on their leisure.

@miekg sorry for the delay and thank you for your patience! I am going to work on this now. Since it has been a while, just wanted to confirm this is how we still want it ?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP plugin/et..." ]
>> make etcd3 the new etcd and add a directive version VERSION that defaults to version 2, but can take a 3 as well; for 2 we use etcdv2 for 3 we use etcdv3 - this is backwards compatible and gives people to upgrade to 3 on their leisure.
>
>@miekg sorry for the delay and thanks for your patience! I am going to work on this now. Since it has been a while, just wanted to confirm this is how we still want it ?

Cool! And yes, this is the (proposed) way forward as I haven't seen any other 
proposals.

**Comment by miekg** [comment]
@nitisht this is ready for review again, isn't? We then need to release a 1.2.0 which deprecates some stuff and then work on merging this and dropping the current etcd2 plugin.

I don't see anything massive that is holding up a 1.2.0 release so we can do this pretty soon.

**Comment by nitisht** [comment]
hey @miekg I have the changes ready as per the approach we discussed, need to test it still though. okay to do this in a couple of days?

**Comment by miekg** [comment]
Sgtm. Thanks!

**Comment by nitisht** [comment]
@miekg I justed tested locally and looks like it is working fine now

**Comment by nitisht** [comment]
@miekg is there something else I can do here to get this in?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/etcdv3..." ]
>@miekg is there something else I can do here to get this in?

No I should review and release 1.2.0 because we can't merge this without 
breakage. Both should happen soon, 1.2.0 I can probably release this evening.

**Comment by nitisht** [comment]
Great, thanks for the info

File: plugin/etcd3/etcd.go
Comment (miekg) [inline plugin/etcd3/etcd.go]: This can be simpllified

Comment (miekg) [reply plugin/etcd3/etcd.go]: return err.Errror() == keyNotfoundErr

File: plugin/etcd3/etcd.go
Comment (miekg) [inline plugin/etcd3/etcd.go]: Why is the `[:]` need here?

File: plugin/etcd3/handler.go
Comment (miekg) [inline plugin/etcd3/handler.go]: compress is done separately now, please remove.

File: plugin/etcd3/lookup_test.go
Comment (miekg) [inline plugin/etcd3/lookup_test.go]: why are these tests removed?

**Review by miekg — COMMENTED** [review]
looks good. I think we should just name it etcd as we're planning to replace it wholesale.

File: plugin/etcd3/lookup_test.go
Comment (nitisht) [reply plugin/etcd3/lookup_test.go]: thanks for the review, looks like there were few tests not running due to renaming. Now that I switched back to the plugin name `etcd` few tests are failing. I will try to fix these today hopefully

**Comment by nitisht** [comment]
@miekg other failing tests are fixed now, only failing test is `TestOtherLookup` https://travis-ci.org/coredns/coredns/jobs/397180376#L898

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/etcdv3..." ]
>@miekg other failing tests are fixed now, only failing test is `TestOtherLookup` https://travis-ci.org/coredns/coredns/jobs/397180376#L898

Ah nice. I should have time this evening (UTC) to help debug this.

**Comment by nitisht** [comment]
> Ah nice. I should have time this evening (UTC) to help debug this.

Awesome, thanks!

**Comment by miekg** [comment]
The failing test is:
~~~
--- FAIL: TestOtherLookup (0.12s)
	helpers.go:131: Additional for "a.mail.skydns.test." contained 10 results, 2 expected
	helpers.go:301: ;; opcode: QUERY, status: NOERROR, id: 24903
		;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 10
		
		;; QUESTION SECTION:
		;a.mail.skydns.test.	IN	 MX
		
		;; ANSWER SECTION:
		a.mail.skydns.test.	300	IN	MX	50 mx.skydns.test.
		
		;; ADDITIONAL SECTION:
		a.ipaddr.skydns.test.	300	IN	A	172.16.1.1
		a.ipaddr.skydns.test.	300	IN	A	172.16.1.1
		a.ipaddr.skydns.test.	300	IN	A	172.16.1.1
		a.ipaddr.skydns.test.	300	IN	A	172.16.1.1
		b.ipaddr.skydns.test.	300	IN	A	172.16.1.2
		mx.skydns.test.	300	IN	CNAME	a.ipaddr.skydns.test.
		mx.skydns.test.	300	IN	CNAME	a.ipaddr.skydns.test.
		mx.skydns.test.	300	IN	CNAME	a.ipaddr.skydns.test.
		mx.skydns.test.	300	IN	CNAME	a.ipaddr.skydns.test.
		mx.skydns.test.	300	IN	CNAME	b.ipaddr.skydns.test.
~~~
If I look at `OtherServices` for the mx we only have:
```
	{Host: "mx.skydns.test", Priority: 50, Mail: true, Key: "a.mail.skydns.test."},
	{Host: "mx.miek.nl", Priority: 50, Mail: true, Key: "b.mail.skydns.test."},
	{Host: "a.ipaddr.skydns.test", Priority: 30, Mail: true, Key: "a.mx.skydns.test."},
```
So I don't understand why `b.ipaddr.skydns.test` pops up here?

**Comment by nitisht** [comment]
IIUC, this is present as well

```
{Host: "b.ipaddr.skydns.test", Mail: true, Key: "b.mx2.skydns.test."},
```

here: https://github.com/coredns/coredns/blob/master/plugin/etcd/other_test.go#L51

**Comment by miekg** [comment]
but that has `mx2` why is it being picked up?

**Comment by nitisht** [comment]
> but that has mx2 why is it being picked up?

I am not really sure, is there a way to run these tests locally? I can debug and try to find out

**Comment by miekg** [comment]
yes, download the docker container:
docker run -d --net=host --name=etcd quay.io/coreos/etcd:v3.3

and run the tests with : go test -tags etcd

**Comment by nitisht** [comment]
> but that has mx2 why is it being picked up?

the problem is change in the way etcd handles directories. etcdv2 supported directories, but there are only flat keys in etcdv3. 

This means a get for `/skydns/test/skydns/mx` returned everything under `/skydns/test/skydns/mx/` in etcdv2. 

Now with etcdv3, a get for `/skydns/test/skydns/mx` with prefix enabled returns everything matching `/skydns/test/skydns/mx*` - meaning `mx/...`, `mx2/..` and `mx3/..` all match the prefix. 

I am not sure how we should handle it now..I have added a quick fix here https://github.com/coredns/coredns/pull/1702/commits/b05ff150c167aa6d5f02e9ba20d3fb50a8f33695#diff-6d0109286d6cd87bf48c9c87a938d4f5R90 but this probably is not the right way..

**Comment by nitisht** [comment]
FWIW this is no clear response filtering in etcdv3 so we could filter out irrelevant responses in get [[ref](https://github.com/coreos/etcd/issues/8467)]

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/etcdv3..." ]
>> but that has mx2 why is it being picked up?
>
>the problem is change in the way etcd handles directories. etcdv2 supported directories, but there are only flat keys in etcdv3.
>
>This means a get for `/skydns/test/skydns/mx` returned everything under `/skydns/test/skydns/mx/` in etcdv2.
>
>Now with etcdv3, a get for `/skydns/test/skydns/mx` with prefix enabled returns everything matching `/skydns/test/skydns/mx*` - meaning `mx/...`, `mx2/..` and `mx3/..` etc match the prefix.

Agh, yes now that makes sense. Thanks for digging into this.

>I am not sure how we should handle it now..I have added a quick fix here https://github.com/coredns/coredns/pull/1702/commits/b05ff150c167aa6d5f02e9ba20d3fb50a8f33695#diff-6d0109286d6cd87bf48c9c87a938d4f5R90 but this probably is not the right way..

No, this needs to be more generic. Can we just always the closing / ?
I'm assuming that will fail, but only for leaves with nothing below, this would 
require two queries though :(

/Miek

--
Miek Gieben

**Comment by nitisht** [comment]
> Can we just always the closing / ?
I'm assuming that will fail, but only for leaves with nothing below, this would
require two queries though :(

Done

**Comment by miekg** [comment]
merged! :-)

Can you send a follow up PR which documents this 2-lookups bit?

**Comment by m4r10k** [comment]
Wow! Really cool!

**Comment by yongtang** [comment]
👍 🎉 🎈 

FYI @JiachengXu

**Comment by jiachengxu** [comment]
Cool!

**Comment by harshavardhana** [comment]
Very cool thanks @miekg

**Comment by miekg** [comment]
@harshavardhana credit where credit is due. I clicked the merge button, @nitisht did the work!

**Comment by nitisht** [comment]
> Can you send a follow up PR which documents this 2-lookups bit?

~Sure, will do~
Done #1918

**Comment by harshavardhana** [comment]
> @harshavardhana credit where credit is due. I clicked the merge button, @nitisht did the work!

Agreed @miekg - Thanks @nitisht

**Comment by nchrisdk** [comment]
Great work! @nitisht

**Pull request merged**
